design/func-version.md

```diff
@@ -0,0 +1,378 @@
++	Func        string                     `json:"func"`
++	FuncVersion string                     `json:"funcVersion,omitempty"`
++	Name        string                     `json:"name"`
++	ObjectRefs  map[string]ObjectReference `json:"objects,omitempty"`
++	Args        map[string]interface{}     `json:"args"`
+ }
+
+ // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+```
+
+An example blueprint with an action that uses this new field looks like this:
+
+```yaml
+actions:
+  tail:
+    phases:
+    - func: KubeExec
+      funcVersion: v0.1.0
+      name: tail
+      args:
+        namespace: default
+        pod: timer
+        container: timer
+        command:
+        - sh
+        - "-c"
+        - |
+         /bin/tail_datetime.sh
+```
+
+If the `funcVersion` property isn't specified in the blueprint action, Kanister
+will determine the stable version using the `VersionedFunc` interface. If the
+Kanister Function doesn't implement the `VersionedFunc` interface, then Kanister
+will fall back to using the existing mechanism to use the default `v0.0.0`
+version of the Function.
+
+### Version Promotion And Deprecation
+
+The timeframe to promote a version to the stable state is based on the
+implementers' discretion. It is expected that there will be a sufficient time
+window between the introduction of a new version and its promotion to the stable
+state, to allow for community feedback and bug fixes.
+
+To use the new non-stable version, users will have to specify the `funcVersion`
+property in their blueprint phase, as shown in the example above. Otherwise, the
+blueprint will continue to use the existing stable (default) version.
+
+Once a new version is promoted to stable, the existing stable version will be
+demoted with a deprecation time frame of at least two releases. During this
+period, log and event warnings will be generated in Kanister's logs and the
+actionset `status` as indication that the in-use version has been deprecated,
+and will be removed in a future release. Users are responsible for updating
+their blueprints to work with the new stable version.
+
+After the deprecation time frame, the deprecated version will officially fall
+out of support. Any usage of this version will cause the controller to return an
+error and halt the execution. The version number will still be preserved for
+posterity.
+
+### Versions Discovery
+
+The versions of all Kanister Functions can be stored in a version YAML file
+that maps the Function name to the supported versions. Special annotation will
+be used to denote all stable versions. This file can be [embedded][4] in
+the Kanister controller binary during build. The controller uses this file to
+determine the stable version of each Function, whenever it needs that
+information.
+
+To facilitate easy version discovery, the controller exposes an endpoint that
+publishes its supported Kanister Functions versions.
+
+## Test Cases
```

Comment:
Should we also include the documentation changes? If multiple versions of the functions have different arguments, how the document is going to look like?

Reply:
Yes, all supported versions will be included in the documentation. Similar to https://github.com/kanisterio/kanister/issues/1342, we should use a script to automate the functions doc generation.